sst/test/include/framework/registry.hh

```diff
@@ -0,0 +1,102 @@
+#pragma once
+
+#include <functional>
+#include <iostream>
+#include <map>
+#include <memory>
+#include <string>
+#include <vector>
+
+#include "framework/context.hh"
+#include "framework/runner.hh"
+#include "framework/uid.hh"
+
+/**
+ * The Registry class is used register Factory(s) which are able
+ * to create Runner(s) associated with TEST_GROUP(s). Registry is a singleton
+ * and can be only be instantiated once.
+ */
+class Registry {
+ public:
+  using Factory = std::function<std::unique_ptr<Runner>()>;
+  using RegistryMap = std::map<std::string, std::vector<Factory>>;
+
+  /** Method creates and returns the singleton instance of Registry. */
+  static Registry* getInstance() {
+    /** Static instance of the singleton Registry class. */
+    static Registry* instance_;
+    if (instance_ == NULL) {
+      instance_ = new Registry();
+    }
+    return instance_;
+  }
+
+  /**
+   * This method returns derived instance of classes extending the Runner class
+   * as polymorphic parent instances i.e the make_unique function calls the
+   * constructor of classes extending the Group class and returns them as
+   * polymorphic Runner(s) i.e std::make_unique<T extends Group>()
+   */
+  template <typename T>
+  static std::unique_ptr<Runner> createDerived() {
+    return std::make_unique<T>();
+  }
+
+  /**
+   * This method registers a Factory responsible for creating polymorphic
+   * Runner(s) instances of classes extending the Group class.
+   */
+  static bool registerGroup(std::string fname, uint64_t line, std::string gname,
+                            Factory f) {
+    UidRegistry::validateGroupName(gname, fname, line);
+    return addRunner(fname, f);
+  };
+
+  /**
+   * This method returns the singleton instance of RegistryMap used to store
+   * all Factory(s).
+   */
+  static RegistryMap* getMap() {
+    /** Static instance of the map used to store all Factory(s). */
+    static RegistryMap* map;
+    if (!map) {
+      map = new RegistryMap;
+    }
+    return map;
+  }
```

Comment:
Again, I'm confused over the form of this function. Is there a reason why the `Registry` class can't use a public constructor, then have privatly defined `instance_` and `map_` variables with getters?

Reply:
There are 2 reasons for this:
1) A static variable map is required because the TEST_GROUP Factory(s) must be registered to the registry to create a test group during runtime. To achieve this, the registration mechanism has to be implemented in a static context, otherwise, we cannot register TEST_GROUP(s) without instantiating them manually i.e by calling the constructor. This would require us to call the constructor of all classes extending `Group` manually (which is undesirable)
2) Since execution in a static context is necessary, we have to deal with something called the static initialisation fiasco. Basically, there is no way to guarantee the initialisation order in a static context, so it could happen that when invoking a static variable via a static function it might not have been initialised (hence leading to a segfault), to ensure that we always access an initialised static variable, the getter pattern is required.

Reply:
Makes sense, thanks for explaining